File: ios/Odeon/Telas/TelaDaTemporada.swift
import SwiftUI

/// Os episódios de uma temporada, no iPhone.
///
/// ⚠️ **Lista, e não grade** — o episódio não se escolhe pela imagem (a capa é a
/// mesma nos dezoito), escolhe-se pelo número e pelo que já aconteceu com ele. A
/// fileira põe quadro, número, título e estado na mesma linha de leitura.
///
/// ⚠️ Falta a sinopse por episódio, e é falta do servidor: `ObraDaLista` não traz
/// `overview`. Ver `PEDIDOS-AO-SERVIDOR.md, «já entregue» 10`. Enquanto não vier, a linha não é
/// desenhada — não se inventa sinopse a partir do título (§18).
struct TelaDaTemporada: View {
    let odeon: RepositorioOdeon
    let modelo: ModeloDaSerie
    let numero: Int
    let aoTocar: (ObraDaLista) -> Void

    private var temporada: TemporadaDaSerie? { modelo.temporada(numero) }

    var body: some View {
        ZStack {
            Cores.fundo.ignoresSafeArea()
            if let t = temporada {
                ScrollView {
                    VStack(alignment: .leading, spacing: 4) {
                        cabecalho(t).padding(.horizontal, 20).padding(.bottom, 14)
                        ForEach(t.episodios) { ep in
                            Fileira(odeon: odeon, episodio: ep) { aoTocar(ep) }
                        }
                    }
                    .padding(.bottom, 28)
                }
            } else {
                Recado(
                    titulo: "esta temporada não está aqui",
                    detalhe: modelo.erro ?? "os episódios dela não chegaram",
                ) { EmptyView() }
            }
        }
        .navigationTitle(temporada?.rotulo ?? "")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func cabecalho(_ t: TemporadaDaSerie) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(t.rotulo).font(Tipo.letreiro(28)).foregroundStyle(Cores.texto)
            Text(sub(t))
                .font(Tipo.rotulo(11))
                .tracking(Tipo.espacoDoRotulo)
                .foregroundStyle(Cores.textoApagado)
        }
    }

    private func sub(_ t: TemporadaDaSerie) -> String {
        var partes = ["\(t.quantos) episódio" + (t.quantos > 1 ? "s" : "")]
        if t.vistos > 0 { partes.append("\(t.vistos) visto" + (t.vistos > 1 ? "s" : "")) }
        return partes.joined(separator: "  ·  ")
    }
}

private struct Fileira: View {
    let odeon: RepositorioOdeon
    let episodio: ObraDaLista
    let aoTocar: () -> Void

    private var visto: Bool { episodio.finished == true }

    private var andado: Double {
        guard let onde = episodio.ondeParou,
              let total = episodio.duracaoEmSegundos, total > 0 else { return 0 }
        return min(max(onde / total, 0), 1)
    }

    var body: some View {
        Button(action: aoTocar) {
            HStack(alignment: .center, spacing: 14) {
                ZStack(alignment: .topTrailing) {
                    /// Mesma armadilha do cartão da temporada: `Color.clear` com
                    /// a arte no `background`, pra a fileira não herdar a largura
                    /// intrínseca da imagem.
                    Color.clear
                        .frame(width: 128)
                        .aspectRatio(16.0 / 9.0, contentMode: .fit)
                        .background { ArteDoOdeon(odeon: odeon, caminho: episodio.arte) }
                        .background(Cores.fundoElevado)
                        .clipShape(.rect(cornerRadius: 8))
                        .overlay(alignment: .bottomLeading) {
                            /// Um **ou** outro: quem terminou não parou no meio.
                            if !visto, andado > 0 { BarraDeAndado(fracao: andado) }
                        }
                        .overlay {
                            if visto { Color.black.opacity(0.45).clipShape(.rect(cornerRadius: 8)) }
                        }
                    if visto { MarcaDeVisto().padding(4) }
                }

                VStack(alignment: .leading, spacing: 4) {
                    HStack(alignment: .top, spacing: 10) {
                        if let n = episodio.episodio {
                            Text("\(n)")
                                .font(Tipo.letreiro(20))
                                .foregroundStyle(Cores.destaqueApagado)
                        }
                        Text(episodio.title)
                            .font(.subheadline.weight(.semibold))
                            /// Visto apaga o título — é a marca que se lê de relance.
                            .foregroundStyle(visto ? Cores.textoApagado : Cores.texto)
                            .lineLimit(2)
                            .multilineTextAlignment(.leading)
                    }
                    Text(sub)
                        .font(Tipo.rotulo(11))
                        .tracking(Tipo.espacoDoRotulo)
                        .foregroundStyle(andado > 0 && !visto ? Cores.destaque : Cores.textoApagado)
                    /// ⚠️ A sinopse chegou em 18/08/2026 — era «falta do
                    /// servidor, não desenho», e deixou de ser. Metade dos
                    /// episódios não tem, e aí a linha **não existe** (§24).
                    if let sinopse = episodio.overview, !sinopse.isEmpty {
                        Text(sinopse)
                            .font(.caption)
                            .foregroundStyle(Cores.textoApagado)
                            .lineLimit(2)
                            .multilineTextAlignment(.leading)
                    }
                }
                Spacer(minLength: 0)
            }
            /// ⚠️ 64pt de mínima e a fileira **inteira** é o alvo — a régua de
            /// toque da casa.
            .frame(minHeight: 64)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .contentShape(.rect)
        }
        .buttonStyle(.plain)
    }

    /// `S01E04 · 43min`, e `· faltam 21min` quando há onde voltar. Cada pedaço só
    /// entra se existir — duração nula não vira `0min` (§18).
    private var sub: String {
        var partes: [String] = []
        if let c = episodio.codigo { partes.append(c) }
        if let d = episodio.duracaoEmSegundos, d > 0 { partes.append("\(Int(d / 60))min") }
        if andado > 0, !visto {
            let faltam = Int(((episodio.duracaoEmSegundos ?? 0) - (episodio.ondeParou ?? 0)) / 60)
            if faltam > 0 { partes.append("faltam \(faltam)min") }
        }
        return partes.joined(separator: "  ·  ")
    }
}

File: android/app/src/main/kotlin/dev/odeon/android/ui/serie/TelaDaTemporada.kt
package dev.odeon.android.ui.serie

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import coil3.compose.AsyncImage
import dev.odeon.android.dados.ObraDaLista
import dev.odeon.android.ui.Cores
import dev.odeon.android.ui.Recado
import dev.odeon.android.ui.Tipo

/// Os episódios de uma temporada, no celular.
///
/// ⚠️ **Lista, como na sala** — e aqui a razão é ainda mais direta: numa largura
/// de celular, uma grade de quadros 16:9 caberia em duas colunas de miniatura
/// ilegível, ou numa coluna de cartões gigantes. A fileira horizontal põe quadro,
/// número, título e estado na mesma linha de leitura, e cabem seis na tela.
///
/// ⚠️ Falta a sinopse por episódio, e é falta do servidor — `ObraDaLista` não
/// traz `overview`. Ver `PEDIDOS-AO-SERVIDOR.md, «já entregue» 10`.
@Composable
fun TelaDaTemporada(
    modelo: ModeloDaSerie,
    numeroDaTemporada: Int,
    aoTocar: (episodioId: String) -> Unit,
    aoVoltar: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val estado by modelo.estado.collectAsStateWithLifecycle()
    val temporada = estado.temporada(numeroDaTemporada)

    if (temporada == null) {
        Recado(
            titulo = "esta temporada não está aqui",
            detalhe = estado.erro ?: "os episódios dela não chegaram",
            aoTentar = modelo::carregar,
            aoVoltar = aoVoltar,
            modifier = modifier,
        )
        return
    }

    LazyColumn(
        modifier.fillMaxSize().background(Cores.fundo),
        contentPadding = PaddingValues(bottom = 28.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        item {
            Column(Modifier.padding(start = 8.dp, end = 20.dp, top = 8.dp, bottom = 14.dp)) {
                TextButton(onClick = aoVoltar) {
                    Text("‹ ${estado.titulo}", color = Cores.destaque, maxLines = 1)
                }
                Spacer(Modifier.height(6.dp))
                Text(
                    text = temporada.rotulo,
                    style = MaterialTheme.typography.headlineSmall,
                    color = Cores.texto,
                    modifier = Modifier.padding(start = 12.dp),
                )
                Spacer(Modifier.height(6.dp))
                Text(
                    text = buildList {
                        add("${temporada.quantos} episódio" + if (temporada.quantos > 1) "s" else "")
                        if (temporada.vistos > 0) {
                            add("${temporada.vistos} visto" + if (temporada.vistos > 1) "s" else "")
                        }
                    }.joinToString("  ·  "),
                    style = Tipo.rotulo,
                    color = Cores.textoApagado,
                    modifier = Modifier.padding(start = 12.dp),
                )
            }
        }

        items(temporada.episodios, key = { it.id }) { ep ->
            FileiraDoEpisodio(
                episodio = ep,
                arte = modelo.arte(ep.arte),
                aoTocar = { aoTocar(ep.id) },
            )
        }
    }
}

@Composable
private fun FileiraDoEpisodio(
    episodio: ObraDaLista,
    arte: String?,
    aoTocar: () -> Unit,
) {
    val visto = episodio.finished == true
    val andado = andadoDoEpisodio(episodio)

    Row(
        Modifier
            .fillMaxWidth()
            /// ⚠️ 64dp de altura mínima e a fileira **inteira** é o alvo — a
            /// régua de toque da casa. Um quadro de 112dp já passa disso, mas a
            /// mínima protege o caso sem arte.
            .heightIn(min = 64.dp)
            .clickable(onClick = aoTocar)
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            Modifier
                .width(128.dp)
                .aspectRatio(16f / 9f)
                .clip(RoundedCornerShape(8.dp)),
        ) {
            if (arte != null) {
                AsyncImage(
                    model = arte,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.fillMaxSize(),
                )
            } else {
                Box(Modifier.fillMaxSize().background(Cores.fundoElevado))
            }
            /// Um **ou** outro — quem terminou não parou no meio.
            if (visto) {
                Box(Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.45f)))
                MarcaDeVisto(Modifier.align(Alignment.TopEnd).padding(4.dp))
            } else if (andado > 0f) {
                BarraDeAndado(andado, Modifier.align(Alignment.BottomStart))
            }
        }

        Spacer(Modifier.width(14.dp))

        Column(Modifier.fillMaxWidth()) {
            Row(verticalAlignment = Alignment.Top) {
                episodio.episodio?.let { n ->
                    Text(
                        text = "$n",
                        style = MaterialTheme.typography.titleMedium,
                        color = Cores.destaqueApagado,
                    )
                    Spacer(Modifier.width(10.dp))
                }
                Text(
                    text = episodio.title,
                    style = MaterialTheme.typography.titleSmall,
                    /// Visto apaga o título — é a marca que se lê de relance.
                    color = if (visto) Cores.textoApagado else Cores.texto,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            Spacer(Modifier.height(4.dp))
            Text(
                text = subtitulo(episodio, andado),
                style = Tipo.rotulo,
                color = if (andado > 0f && !visto) Cores.destaque else Cores.textoApagado,
            )
            /// ⚠️ A sinopse do episódio — chegou em 18/08/2026. Metade não tem, e
            /// aí a linha **não existe** (§24).
            episodio.overview?.takeIf { it.isNotBlank() }?.let { sinopse ->
                Spacer(Modifier.height(4.dp))
                Text(
                    text = sinopse,
                    style = MaterialTheme.typography.bodySmall,
                    color = Cores.textoApagado,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

/// `S01E04 · 43min`, e `· faltam 21min` quando há onde voltar.
///
/// ⚠️ Cada pedaço só entra se existir — duração nula não vira `0min` (§18).
private fun subtitulo(ep: ObraDaLista, andado: Float): String = buildList {
    ep.codigo?.let { add(it) }
    ep.duracaoEmSegundos?.takeIf { it > 0 }?.let { add("${(it / 60).toInt()}min") }
    if (andado > 0f && ep.finished != true) {
        val faltam = (((ep.duracaoEmSegundos ?: 0.0) - (ep.ondeParou ?: 0.0)) / 60).toInt()
        if (faltam > 0) add("faltam ${faltam}min")
    }
}.joinToString("  ·  ")

private fun andadoDoEpisodio(ep: ObraDaLista): Float {
    val onde = ep.ondeParou ?: return 0f
    val total = ep.duracaoEmSegundos?.takeIf { it > 0 } ?: return 0f
    return (onde / total).toFloat().coerceIn(0f, 1f)
}
